### PR TITLE
[BUG] #655 - Some special characters truncate the serialized data in db

### PR DESCRIPTION
**Pull request by crouillon**

### CODE REVIEW

File: ClassContent/AbstractContent.php [hunk 1]
```diff
@@ -284,9 +284,11 @@ public function __set($var, $value)
 
                 if (is_object($value) && $value instanceof AbstractClassContent) {
                     $value = $this->_addSubcontent($value);
+                } elseif ('scalar' === $type) {
+                    $val[] = array($type => json_encode($value, JSON_ERROR_NONE ));
+                } else {
+                    $val[] = array($type => $value);
                 }
```
Comment (eric-chau) [inline ClassContent/AbstractContent.php]: This part introduces a code regression. $val[] = array($type => $value); is missing cause you move it into the `else` block

File: ClassContent/AbstractContent.php [hunk 2]
```diff
@@ -284,9 +284,11 @@ public function __set($var, $value)
 
                 if (is_object($value) && $value instanceof AbstractClassContent) {
                     $value = $this->_addSubcontent($value);
+                } elseif ('scalar' === $type) {
+                    $val[] = array($type => json_encode($value, JSON_ERROR_NONE ));
```
Comment (eric-chau) [inline ClassContent/AbstractContent.php]: Seems that JSON_ERROR_NONE is not a supported options. Valid ones: `JSON_HEX_QUOT, JSON_HEX_TAG, JSON_HEX_AMP, JSON_HEX_APOS, JSON_NUMERIC_CHECK, JSON_PRETTY_PRINT, JSON_UNESCAPED_SLASHES, JSON_FORCE_OBJECT, JSON_PRESERVE_ZERO_FRACTION, JSON_UNESCAPED_UNICOD`.

`JSON_ERROR_NONE` is one of the possible constant returned by `json_last_error()`.

File: ClassContent/AbstractContent.php [hunk 3]
```diff
@@ -930,7 +932,10 @@ public function getData($var = null, $forceArray = false)
                 $value = end($values);
             }
 
-            if ($type !== 'scalar' && $type !== 'array') {
+            if ('scalar' === $type) {
+                $decode = json_decode($value);
+                $data[] = $decode ?: $value;
```
Comment (eric-chau) [inline ClassContent/AbstractContent.php]: According to comment below, this is not a valid check.

**Comment by crouillon** [comment]
Thanks @eric-chau but it's a little early to code review. I can't test locally for now so i alredy create the PR to get them
Renamed this PR [WIP] till it's ok

**Comment by eric-chau** [comment]
ok @crouillon sorry!

**Comment by eric-chau** [comment]
I will just add some explanations about how @crouillon fixed this issue. The issue is that for some special characters like emoji php `serialize()` is not able to do it correctly so it truncate the serialized string. This will result on an invalid insertion into database. To prevent this we first call `json_encode` on every scalar element then Doctrine can safely pass it to php `serialize()`. Indeed `json_serialize()` is able to convert special characters into valid character for php `serialize()`.

Hope this help you to understand the issue.

:+1:, thank you @crouillon

**Pull request merged**